### PR TITLE
A couple of changes in bpf/nat to help adding new support of ICMP types

### PR DESCRIPTION
**Pull request by sahid**
Refactor some parts of `lib/nat.h` to make easier future support of ICMP message types

### CODE REVIEW

File: bpf/lib/nat.h [hunk 1]
```diff
@@ -445,18 +445,25 @@ static __always_inline int snat_v4_rewrite_ingress(struct __ctx_buff *ctx,
...
+			__u8 type = 0;
 			__be32 from, to;
 
-			if (ctx_store_bytes(ctx, off +
-					    offsetof(struct icmphdr, un.echo.id),
-					    &state->to_dport,
-					    sizeof(state->to_dport), 0) < 0)
+			if (ctx_load_bytes(ctx, off +
+					   offsetof(struct icmphdr, type),
+					   &type, 1) < 0)
```
Comment (aspsk) [inline bpf/lib/nat.h:453]: We should return `DROP_INVALID` when `ctx_load_bytes` fails

File: bpf/lib/nat.h [hunk 2]
```diff
@@ -831,15 +831,17 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
 	case IPPROTO_ICMP:
```
Comment (aspsk) [inline bpf/lib/nat.h:838]: A typo in the commit message: `convenientchange`

**Review by aspsk — CHANGES_REQUESTED** [review]

File: bpf/lib/nat.h (hunk 1 above)
Comment (sahid) [reply bpf/lib/nat.h:453]: Oh right I already received comment for this exact same point. Sorry will fix it. Thank you.

**Review by aspsk — APPROVED** [review]

**Comment by sahid** [comment]
BPF check test failing because of #22043. I will rebase after the change is merged.

**Comment by sahid** [comment]
o/ @joestringer just in case you have not noticed it. I have addressed you request to separate some commits from #18414

File: bpf/lib/nat.h [hunk 3]
```diff
@@ -445,18 +445,25 @@ static __always_inline int snat_v4_rewrite_ingress(struct __ctx_buff *ctx,
...
+						    offsetof(struct icmphdr, un.echo.id),
+						    &state->to_dport,
+						    sizeof(state->to_dport), 0) < 0)
+					return DROP_WRITE_ERROR;
+				from = tuple->dport;
+				to = state->to_dport;
+				flags = 0; /* ICMPv4 has no pseudo-header */
+				sum_l4 = csum_diff(&from, 4, &to, 4, 0);
+				csum.offset = offsetof(struct icmphdr, checksum);
+			}
```
Comment (pchaigno) [inline bpf/lib/nat.h:466]: Unless I'm missing something, this is a functional change. What's the impact for users? Does it change how the echo/echo-reply messages are treated? What will be the visible change?

File: bpf/lib/nat.h [hunk 4]
```diff
@@ -1447,19 +1447,21 @@ snat_v6_rev_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
...
 			tuple.dport = 0;
 			tuple.sport = icmp6hdr.icmp6_dataun.u_echo.identifier;
-		} else {
+			break;
+		case ICMPV6_ECHO_REPLY:
 			tuple.dport = icmp6hdr.icmp6_dataun.u_echo.identifier;
 			tuple.sport = 0;
+			break;
+		default:
+			return DROP_NAT_UNSUPP_PROTO;
```
Comment (pchaigno) [inline bpf/lib/nat.h:1476]: The first two commits could be merge together and the title changed to `bpf/nat: refactor ICMP handlers into switch`

File: bpf/lib/nat.h (hunk 3 above)
Comment (sahid) [reply bpf/lib/nat.h:466]: Since we do currently only support echo/echo-reply no changes are expected from an user POV.

File: bpf/lib/nat.h (hunk 4 above)
Comment (sahid) [reply bpf/lib/nat.h:1476]: Sure, done.

File: bpf/lib/nat.h (hunk 3 above)
Comment (pchaigno) [reply bpf/lib/nat.h:466]: > we do currently only support echo/echo-reply

What does this mean exactly? Do you mean in the documentation? If someone tries to send something other than echo/echo-reply what happens? Or do you mean that the information collected here is dropped later on if the packet is not an ICMP echo or echo-reply (in which case this is *not* a functional change)?

Comment (sahid) [reply bpf/lib/nat.h:466]: Oh, yes yes there are no functional changes today as the packet will be dropped before to join this function [0]. The point will be once we will support more types.

[0] https://github.com/cilium/cilium/pull/22004/files#diff-ef710180f228e8f6bbad249e2670dbc42a54f45da3bbc4c5aebbb1f3865979e9R779

Comment (pchaigno) [reply bpf/lib/nat.h:466]: Ok, I see.

Could you update the commit description to make that clear? Right now the commit title sounds like this commit is limiting the rewrite of ICMP messages on ingress when it's not changing anything.

I think you can also merge the last two commits in one, though it doesn't matter much.

Comment (sahid) [reply bpf/lib/nat.h:466]: Yes, done.

**Review by pchaigno — APPROVED** [review]

**Comment by sahid** [comment]
@pchaigno do you see why the CI is failing? I'm not able to understand the reason when looking into Jenkins

**Comment by pchaigno** [comment]
/test-1.16-4.9

**Comment by sahid** [comment]
Thank you for the quick merge Tobias.

**Pull request merged**